### PR TITLE
emit hooks from a view as long as the path is inside the default root

### PR DESCRIPTION
**Pull request by icewind1991**
For https://github.com/owncloud/files_drop/issues/61

cc @DeepDiver1975 @PVince81

### CODE REVIEW

**Comment by PVince81** [comment]
This is likely to have other consequences. If parts of OC are using a root view and were not expecting hooks, but now we fired them. Things could go wrong.

I'd rather have this in 8.2 for proper regression testing.
Is it possible to fix it on the other side and have the view use the proper path short term ?

File: tests/lib/files/view.php [hunk 1]
```diff
@@ -1080,4 +1080,33 @@ public function testGetRelativePathOnNull() {
 	public function testNullAsRoot() {
 		new \OC\Files\View(null);
 	}
+
+	public function hookPathProvider() {
+		return [
+			['/foo/files', '/foo', true],
+			['/foo/files/bar', '/foo', true],
+			['/foo', '/foo', false],
+			['/foo', '/files/foo', true],
```
Comment (PVince81) [inline tests/lib/files/view.php]: please add "files/foo" too (without leading slash), we know that many pieces of the code will break if it is.

**Comment by DeepDiver1975** [comment]
This change fixes owncloud/files_drop#61 and owncloud/files_drop#63 :+1:

**Comment by DeepDiver1975** [comment]
> This is likely to have other consequences. If parts of OC are using a root view and were not expecting hooks, but now we fired them. Things could go wrong.

This is indeed something to have a look at ... because of  owncloud/files_drop#63 this has an backporting impact :boom:

**Comment by sbelov1** [comment]
Hey guys,

 Is this ready to provide as a short term fix for the prospect? Would it take care of both owncloud/files_drop#61 and owncloud/files_drop#63?

**Comment by DeepDiver1975** [comment]
> Hey guys,
> 
> Is this ready to provide as a short term fix for the prospect? Would it take care of both owncloud/files_drop#61 and owncloud/files_drop#63?

No

**Comment by icewind1991** [comment]
Do we want this for 8.1 or 8.2?

We can do https://github.com/owncloud/files_drop/pull/62 for the other versions

**Comment by DeepDiver1975** [comment]
> Do we want this for 8.1 or 8.2?

well - the behavior is broken back until oc7 - sooner or later we need to fix this in all stable branches.

Question is if we can limit the risk as noted by @PVince81

**Comment by DeepDiver1975** [comment]
@icewind1991 please rebase

**Comment by DeepDiver1975** [comment]
@PVince81 @icewind1991 is there a chance to reduce the risk as declared above? Maybe we can enable/disable hook processing using an additional ctor parameter? 

Because this issue also affects oc8 we shall put some traction on this - THX

**Comment by PVince81** [comment]
Or another alternative / quickfix would be change the "broken" view instance that does not fire hook and instantiate it as a user view instead of root view. This way the risk is limited to that specific code path.

And keep this PR for 8.2. @icewind1991 is that possible ?

**Comment by icewind1991** [comment]
> Or another alternative / quickfix would be change the "broken" view instance that does not fire hook and instantiate it as a user view instead of root view. This way the risk is limited to that specific code path.

That would require even more changes

> is there a chance to reduce the risk as declared above? Maybe we can enable/disable hook processing using an additional ctor parameter?

That wouldn't really help since the view that caused the issue is the one used by the Node api which is the main view which would be effected by this change

https://github.com/owncloud/files_drop/pull/62

**Comment by DeepDiver1975** [comment]
> A work around like owncloud/files_drop#62 would be the only way to fix the problem with low risk.

this is a rather damn dirty hack - I'd prefer not to merge  owncloud/files_drop#62 - at least not this way.

hmmm ... can we elaborate on the risk merging this for 8.1?

**Comment by icewind1991** [comment]
I dont think the risk is to high, even if there are hooks emitted I dont think there's any negative effect a hook can have besides possible taking some server resources

**Comment by DeepDiver1975** [comment]
> I dont think the risk is to high, even if there are hooks emitted I dont think there's any negative effect a hook can have besides possible taking some server resources

@PVince81 your call

**Comment by PVince81** [comment]
If we assume that our hooks listeners are "sane" and do check in what folder they are working, then the risk is low. If not, then it might happen that the versions app creates versions for files outside of "files" or things like that. But as far as I remember there should be checks already.

Some regression testing will need to be scheduled though.

**Comment by PVince81** [comment]
:+1:  makes sense

**Comment by icewind1991** [comment]
> it might happen that the versions app creates versions for files outside of "files"

It still only emits hooks for files inside `/$user/files`

**Comment by DeepDiver1975** [comment]
> I dont think the risk is to high, even if there are hooks emitted I dont think there's any negative effect a hook can have besides possible taking some server resources

@icewind1991 can I ask you to perform some tests (especially version handling) so that we are fully sure that thing will not fall apart? THX

**Comment by icewind1991** [comment]
Found another issue is determine whether hooks should be triggered.

Version handling now works as expected, both trough the web interface, webdav and apps using the Node api (tested with Notes)

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
let's move this in - fixes serious issues - remaining test should show regressions ....

**Comment by MorrisJobke** [comment]
@icewind1991 @karlitschek Is the backport okay?

**Comment by karlitschek** [comment]
backport is fine after the next back port release is out of the door.

**Comment by MorrisJobke** [comment]
@icewind1991 Please open the backport PR :)

**Comment by icewind1991** [comment]
Backport is here: #17172

**Pull request merged**